packages/backend/src/server/api/ApiCallService.ts

```diff
@@ -44,92 +50,144 @@ export class ApiCallService implements OnApplicationShutdown {
-						ctx.set('Cache-Control', `public, max-age=${endpoint.meta.cacheSec}`);
-					}
-					reply(res);
-				}).catch((e: ApiError) => {
-					reply(e.httpStatusCode ? e.httpStatusCode : e.kind === 'client' ? 400 : 500, e);
-				});
-		
-				// Log IP
-				if (user) {
-					this.metaService.fetch().then(meta => {
-						if (!meta.enableIpLogging) return;
-						const ip = ctx.ip;
-						const ips = this.userIpHistories.get(user.id);
-						if (ips == null || !ips.has(ip)) {
-							if (ips == null) {
-								this.userIpHistories.set(user.id, new Set([ip]));
-							} else {
-								ips.add(ip);
-							}
-		
-							try {
-								this.userIpsRepository.createQueryBuilder().insert().values({
-									createdAt: new Date(),
-									userId: user.id,
-									ip: ip,
-								}).orIgnore(true).execute();
-							} catch {
-							}
-						}
-					});
-				}
-			}).catch(e => {
-				if (e instanceof AuthenticationError) {
-					reply(403, new ApiError({
-						message: 'Authentication failed. Please ensure your token is correct.',
-						code: 'AUTHENTICATION_FAILED',
-						id: 'b0a7f5f8-dc2f-4171-b91f-de88ad238e14',
-					}));
-				} else {
-					reply(500, new ApiError());
+	public handleRequest(
+		endpoint: IEndpoint & { exec: any },
+		request: FastifyRequest<{ Body: Record<string, unknown>, Querystring: Record<string, unknown> }>,
+		reply: FastifyReply,
+	) {
+		const body = request.method === 'GET'
+			? request.query
+			: request.body;
+
+		const token = body['i'];
+		if (token != null && typeof token !== 'string') {
+			reply.code(400);
+			return;
+		}
+		this.authenticateService.authenticate(token).then(([user, app]) => {
+			this.call(endpoint, user, app, body, null, request).then((res: any) => {
+				if (request.method === 'GET' && endpoint.meta.cacheSec && !body['i'] && !user) {
+					reply.header('Cache-Control', `public, max-age=${endpoint.meta.cacheSec}`);
 				}
+				this.send(reply, res);
+			}).catch((err: ApiError) => {
+				this.send(reply, err.httpStatusCode ? err.httpStatusCode : err.kind === 'client' ? 400 : 500, err);
 			});
+
+			if (user) {
+				this.logIp(request, user);
+			}
+		}).catch(err => {
+			if (err instanceof AuthenticationError) {
+				this.send(reply, 403, new ApiError({
+					message: 'Authentication failed. Please ensure your token is correct.',
+					code: 'AUTHENTICATION_FAILED',
+					id: 'b0a7f5f8-dc2f-4171-b91f-de88ad238e14',
+				}));
+			} else {
+				this.send(reply, 500, new ApiError());
+			}
 		});
 	}
 
+	public async handleMultipartRequest(
+		endpoint: IEndpoint & { exec: any },
+		request: FastifyRequest<{ Body: Record<string, unknown>, Querystring: Record<string, unknown> }>,
+		reply: FastifyReply,
+	) {
+		const multipartData = await request.file();
+		if (multipartData == null) {
+			reply.code(400);
+			return;
+		}
+
+		const [path] = await createTemp();
+		await pump(multipartData.file, fs.createWriteStream(path));
+	
+		const token = multipartData.fields['i'];
```

Comment:
The problem is that the `multipartData.fields` property returns an object like this:
```js
{
  i: {
    fieldname: 'i',
    mimetype: 'text/plain',
    value: /* ... */,
    // ...
  },
  force: {
    fieldname: 'force',
    mimetype: 'text/plain',
    value: 'true',
    // ...
  },
  file: {
    fieldname: 'file',
    filename: 'blob',
    mimetype: /* ... */,
    file: /* ... */,
    // ...
  },
  name: {
    fieldname: 'name',
    mimetype: 'text/plain',
    value: /* ... */,
    // ...
  }
}
```
The `extractMultipartFields` method converts this object into the following format:
```js
{
  i: /* ... */,
  force: 'true',
  file: undefined,
  name: /* ... */
}
```
However, one of the problems of the `extractMultipartFields` method is that it contains `file: undefined`, which does not exist on [koa's equivalent](https://github.com/misskey-dev/misskey/blob/01d4d55e78fd977b9d44a68a2504e6091d346b7a/packages/backend/src/server/api/ApiCallService.ts#L49). I'm not sure if this would be a problem though.